src/lib/kadm5/srv/svr_principal.c

```diff
@@ -405,14 +405,13 @@ kadm5_create_principal_3(void *server_handle,
         kdb->expiration = handle->params.expiration;
 
     kdb->pw_expiration = 0;
-    if (have_polent) {
-        if(polent.pw_max_life)
+    if ((mask & KADM5_PW_EXPIRATION)) {
+        kdb->pw_expiration = entry->pw_expiration;
+    } else if (have_polent) {
+        mask |= KADM5_PW_EXPIRATION;
+        if (polent.pw_max_life)
             kdb->pw_expiration = ts_incr(now, polent.pw_max_life);
-        else
-            kdb->pw_expiration = 0;
```

Comment:
I don't think we want to set the mask bit if polent.pw_max_life is 0, so line 410 could be written `} else if (have_polent && polent.pw_max_life)`.

I am concerned about modifying the value of mask, since we currently don't do that (we set kdb->mask to mask plus the bits we want to set) and it muddies the semantics of the variable.  It might be better to set kdb->mask earlier in the function, and then add to it here.

Reply:
I think we do - actually this is one of my test cases.  If the policy is updated to remove expiration limits, we'd like that to propgate, I think.

Reply:
(Stopped modifying `mask`, though.)

Reply:
For this particular block we're creating a new new principal, though; there is no need to set the mask bit if we don't want pw_expiration set.

Reply:
Updated.